bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/client/MqttAsyncClientWrapper.java

```diff
@@ -35,7 +35,7 @@
 @NonNullByDefault
 public abstract class MqttAsyncClientWrapper {
     private final Logger logger = LoggerFactory.getLogger(MqttAsyncClientWrapper.class);
-    private final List<String> subscribedTopics = new ArrayList<>();
+    private final Set<ClientCallback> subscriptions = new CopyOnWriteArraySet<>();
```

Comment:
I am not sure, if CopyOnWriteArraySet is the best implementation choice here.
All the operations we are doing are add and remove, which are the expensive ones.
I would have gone for <code>ConcurrentHashMap.newKeySet()</code>.

Reply:
good point.